mockgcp/mockcontainer/nodepool.go

```diff
@@ -0,0 +1,209 @@
+// Copyright ccc Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package mockcontainer
+
+import (
+	"context"
+	"strings"
+
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"google.golang.org/protobuf/encoding/prototext"
+	"google.golang.org/protobuf/proto"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/klog/v2"
+
+	"github.com/GoogleCloudPlatform/k8s-config-connector/mockgcp/common/projects"
+	pb "github.com/GoogleCloudPlatform/k8s-config-connector/mockgcp/generated/mockgcp/container/v1beta1"
+)
+
+type nodePoolName struct {
+	Project  *projects.ProjectData
+	Location string
+	Cluster  string
+	NodePool string
+}
+
+func (n *nodePoolName) String() string {
+	return "projects/" + n.Project.ID + "/locations/" + n.Location + "/clusters/" + n.Cluster + "/nodePools/" + n.NodePool
+}
+
+func (n *nodePoolName) ClusterName() *clusterName {
+	return &clusterName{
+		Project:  n.Project,
+		Location: n.Location,
+		Cluster:  n.Cluster,
+	}
+}
+
+// parseNodePoolName parses a string into a nodePoolName.
+// The expected form is `projects/*/locations/*/clusters/*`.
+func (s *MockService) parseNodePoolName(name string) (*nodePoolName, error) {
+	tokens := strings.Split(name, "/")
+
+	if len(tokens) == 8 && tokens[0] == "projects" && tokens[2] == "locations" && tokens[4] == "clusters" && tokens[6] == "nodePools" {
+		project, err := s.projects.GetProjectByID(tokens[1])
```

Comment:
Would it make sense to extract a utility for getting the project from a string with "projects/ID" ?

Reply:
GetProjectByID is already returning a project object.  This code is pretty boilerplate, and I have been thinking it would be nice to be able to pass in a "template" like `projects/:project/locations/:location/clusters/:cluster/nodePools/:nodePool` and have it automatically parsed.  I haven't done that yet though, in practice Duet just does most of the work for these methods for us automatically